src/game/boe.fileio.cpp

```diff
@@ -93,6 +93,9 @@ void finish_load_party(){
 	
 	overall_mode = town_restore ? MODE_TOWN : MODE_OUTDOORS;
 	stat_screen_mode = MODE_INVEN;
+	// build_outdoors call add_outdoor_maps, so we must update them first
+	// checkme: true for actual files, unsure for a legacy saved file
+	save_outdoor_maps();
```

Comment:
I can't imagine any way that this would be correct. In simplified pseudo-code, what this does is:

```cpp
univ.scenario.outdoors.maps = univ.out.out_e;
univ.out.out_e = univ.scenario.outdoors.maps;
```

That clearly doesn't make any sense at all – it's putting garbage in `out_e` and then loading that garbage back in.

If the issue is that `out_e` contained garbage to begin with, then the solution would be to make it not contain garbage.

Thinking on it, I think the correct answer might ultimately be that `out_e` is just obsolete and we delete it entirely. I think its reason for existing was that the original game had such a tiny memory footprint that it couldn't afford to keep the entire scenario in memory all at the same time.

Reply:
Yes this is tricky, I think that when we retrieved a save ( and if we are outdoor) the current map is retrieved by 
```
void cCurOut::readFrom(std::istream& file) {
	readArray(file, expl, 96, 96);
	readArray(file, out, 96, 96);
	readArray(file, out_e, 96, 96);
}
```
so if we do not want to loose it, we need first to retrieve the data of out_e, ... ( which was what save_outdoor_maps do ).

**Note** we need these two maps because originally, when we change sectors, the game rereads the files of the new sector (because it doesn't keep in memory all the maps of the sectors). A similar problem appears with the city, when we enter a city, normally we have to reread the city file (which we don't do). In my new branch I tried to correct it with https://github.com/fosnola/cboe/commit/684929456788b9cba4f90faf98fe7cf83b7a4eeb (and I also add a persistent flag so the coder can decide to have persistent town or town which are reset each time a user enter in a town )

Reply:
> we need two maps because originally, when we change sectors, the game rereads in the scenario the data corresponding to the new sector (because it doesn't keep in memory all the maps of the sectors). A similar problem appears with the city, when we enter a city, normally we have to reread the city file (which we don't do).

Right, but from what I understand, this was just a performance optimization and isn't needed. We could easily save maps for the entire scenario without issue.



> In my new branch I tried to correct it with [fosnola@6849294](https://github.com/fosnola/cboe/commit/684929456788b9cba4f90faf98fe7cf83b7a4eeb) (and I also add a persistent flag so the coder can decide to have persistent town or town which are reset each time a user enter in a town )

That commit doesn't really make much sense. It seems to be overly complicated for what sounds like a very simple feature.

Reply:
> Yes, another (more elegant) solution would be to update univ.scenario.outdoors.maps before storing a game. So when we reload a game, we would not have to rebuild this map using univ.scenario.outdoors.maps (which it is almost ok) and out_e (which contains the current portion)

This does sound better and looks less strange. That would mean calling … `save_outdoor_maps()` just before saving the game?

Reply:
It is probably best to create a function (update_map or something like that) to execute
```
	location corner = univ.party.outdoor_corner;
	for(short i = 0; i < 48; i++)
		for(short j = 0; j < 48; j++) {
			univ.scenario.outdoors[corner.x][corner.y]->maps[j][i] = univ.out.out_e[i][j];
			if(corner.x + 1 < univ.scenario.outdoors.width())
				univ.scenario.outdoors[corner.x + 1][corner.y]->maps[j][i] = univ.out.out_e[i + 48][j];
			if(corner.y + 1 < univ.scenario.outdoors.height())
				univ.scenario.outdoors[corner.x][corner.y + 1]->maps[j][i] = univ.out.out_e[i][j + 48];
			if(corner.y + 1 < univ.scenario.outdoors.height() && corner.x + 1 < univ.scenario.outdoors.width())
				univ.scenario.outdoors[corner.x + 1][corner.y + 1]->maps[j][i] = univ.out.out_e[i + 48][j + 48];
		}
```
Then we can call this function before saving ( and we can modify the first part of save_outdoor_map to call this new function)

Reply:
Huh? But that's the entire content of `save_outdoor_maps()`?

Reply:
Oops, yes, this is the code of save_outdoor_maps ( so you can call save_outdoor_maps ).

**Note:** my code is different because I added the ability to also create cities that keep the terrain changes. For example, if you destroy a building, you can always return to the city and the building is still destroyed...

Reply:
Right, so even after this pull request is finished, our two branches won't even be close to synced…

It's definitely good to deal with the more obvious fixes like this first before deciding what to do with actual new features you've implemented, though.